src/builder/code.rs

```diff
@@ -67,8 +67,8 @@ where
 	}
 
 	/// Add multiple results to signature builder
-	pub fn with_results(mut self, value_types: Vec<elements::ValueType>) -> Self {
-		self.signature.results_mut().extend(value_types);
+	pub fn with_results(mut self, value_types: &[elements::ValueType]) -> Self {
+		self.signature.results_mut().extend(value_types.iter().copied());
```

Comment:
```suggestion
	pub fn with_results<I>(mut self, value_types: &[elements::ValueType]) -> Self
	where
	    I: IntoIterator<Item = elements::ValueType>
	{
		self.signature.results_mut().extend(value_types);
```

Reply:
Certainly possible, though something to consider is the impact of generics on compile times. Having generics in public interfaces seems to be discouraged if it's not really needed. Not sure what the best tradeoff is here.

Reply:
> Having generics in public interfaces seems to be discouraged if it's not really needed.

Discouraged by whom? The Rust API guidelines [say otherwise](https://rust-lang.github.io/api-guidelines/flexibility.html#c-generic). They are **especially** important in public interfaces as we don't know how people want to use them. Given, there is an almost generic type (the slice) that can be used by most. However, what if your elements are not really layed out consecutively in memory? Why do we want to force every caller to `collect` them for no reason? 

I don't think we want to care too much about the downsides of monomorphizations here (compile times, code size) because this is a very small function (its content is already generic). A caller which is cautious about code size can just make sure to always pass the same type as they would need to do anyways with the current code.

Reply:
sure, just thought I'd bring it up. I'll switch things over to generic iterators then